uqcsbot/starboard.py

```diff
@@ -0,0 +1,202 @@
+        reaction = discord.utils.get(recv_message.reactions, emoji=self.starboard_emoji)
+
+        new_reaction_count = 0
+        if reaction is not None:
+            new_reaction_count = reaction.count
+        
+        sb_message_id = self._query_sb_message(recv_message.id)
+
+        if new_reaction_count >= self.base_threshold and sb_message_id is None:
+            new_sb_message = await self.starboard_channel.send(
+                content=f"{str(self.starboard_emoji)} {new_reaction_count} | {recv_message.channel.mention}",
+                embeds=self._create_sb_embed(recv_message)
+                # note that the embed is never edited, which means the content of the starboard post is fixed as soon as the 5th reaction is processed
+            )
+            await new_sb_message.edit(view=discord.ui.View.from_message(new_sb_message).add_item(discord.ui.Button(label="Original Message", style=discord.ButtonStyle.link, url=recv_message.jump_url)))
+
+            self._update_sb_message(recv_message.id, new_sb_message.id)
+        elif new_reaction_count > self.base_threshold and sb_message_id is not None:
+            old_sb_message = await self.starboard_channel.fetch_message(sb_message_id)
+
+            await old_sb_message.edit(content=f"{str(self.starboard_emoji)} {new_reaction_count} | {recv_message.channel.mention}")
+
+            if new_reaction_count >= self.big_threshold and not old_sb_message.pinned:
+                await old_sb_message.pin(reason="Reached 20 starboard reactions")
+
+
+    @commands.Cog.listener()
+    async def on_raw_reaction_remove(self, payload: discord.RawReactionActionEvent):
+        if payload.emoji.id != self.starboard_emoji.id:
+            return
+        
+        guild = self.bot.get_guild(payload.guild_id)
+        if guild is None:
+            return
+        
+        channel = self.bot.get_channel(payload.channel_id)
+        if channel == self.starboard_channel or channel.category.name.startswith("admin"):
+            # TODO: "reaction count" for starboard could take into account (original + starboard)
+            return
+        
+        recv_message = await channel.fetch_message(payload.message_id)
+        reaction = discord.utils.get(recv_message.reactions, emoji=self.starboard_emoji)
+
+        new_reaction_count = 0
+        if reaction is not None:
+            new_reaction_count = reaction.count
+        
+        sb_message_id = self._query_sb_message(recv_message.id)
+        if sb_message_id is None:
+            return
+
+        sb_message = await self.starboard_channel.fetch_message(sb_message_id)
+
+        if new_reaction_count < self.STARBOARD_BASE_THRESHOLD:
+            await sb_message.delete() # delete will also unpin
+            self._remove_sb_message(payload.message_id)
+            return
+        
+        if new_reaction_count < self.STARBOARD_BIG_THRESHOLD and sb_message.pinned:
+            await sb_message.unpin()
```

Comment:
Just an observation, would it be an issue if some pest decided to keep reacting and unreacting a message?

Reply:
Considering you're not the first person to think of this, I may attempt some ratelimiting on the starboard thresholds.

Pinning is less of a concern than the base_threshold sending and unsending the message repeatedly, although I've literally never seen it be a problem in other servers - it only takes one other server user to react or unreact and put the threshold out of reach of the "pest".

Reply:
You raise a good point so I'll leave it to you whether or not there should be a ratelimit

Reply:
added a basic 60-second ratelimit